app/org/sagebionetworks/bridge/upload/IosSchemaValidationHandler.java

```diff
@@ -0,0 +1,559 @@
+        HealthDataRecordBuilder recordBuilder = healthDataDao.getRecordBuilder();
+        context.setHealthDataRecordBuilder(recordBuilder);
+        Map<String, byte[]> attachmentMap = new HashMap<>();
+        context.setAttachmentsByFieldName(attachmentMap);
+
+        // health data records fields
+        recordBuilder.withHealthCode(upload.getHealthCode());
+        recordBuilder.withStudyId(studyId);
+        recordBuilder.withUploadDate(upload.getUploadDate());
+
+        // create an empty object node in our record builder, which we'll fill in as we go
+        ObjectNode dataMap = BridgeObjectMapper.get().createObjectNode();
+        recordBuilder.withData(dataMap);
+
+        JsonNode infoJson = jsonDataMap.get("info.json");
+        if (infoJson == null) {
+            // Recover by replacing this with an empty map
+            addMessageAndWarn(context, String.format("upload ID %s does not contain info.json file", uploadId));
+            infoJson = BridgeObjectMapper.get().createObjectNode();
+
+            // Add it back to the jsonDataMap, since all the logic assumes it contains info.json.
+            jsonDataMap.put("info.json", infoJson);
+        }
+
+        // Use info.json verbatim is the metadata.
+        recordBuilder.withMetadata(infoJson);
+
+        // extract other data from info.json
+        JsonNode taskRunNode = infoJson.get("taskRun");
+        String taskRunId = taskRunNode != null ? taskRunNode.textValue() : null;
+        JsonNode itemNode = infoJson.get("item");
+        String item = itemNode != null ? itemNode.textValue() : null;
+        
+        // Make sure all files specified by info.json are accounted for.
+        // Because ParseJsonHandler moves files from unzippedDataMap to jsonDataMap, there is no overlap between the
+        // two maps.
+        Set<String> fileNameSet = new HashSet<>();
+        fileNameSet.addAll(jsonDataMap.keySet());
+        fileNameSet.addAll(unzippedDataMap.keySet());
+
+        // fileList.size() should be exactly 1 less than fileNameSet.size(), because fileList.size() doesn't include
+        // info.json.
+        JsonNode fileList = infoJson.get("files");
+        if (fileList == null) {
+            // Recover by replacing this with an empty list
+            addMessageAndWarn(context,
+                    String.format("upload ID %s info.json does not contain file list", uploadId));
+            fileList = BridgeObjectMapper.get().createArrayNode();
+        } else if (fileList.size() == 0) {
+            addMessageAndWarn(context, String.format("upload ID %s info.json contains empty file list", uploadId));
+        } else if (fileList.size() != fileNameSet.size() - 1) {
+            addMessageAndWarn(context, String.format("upload ID %s info.json reports %d files, but we found %d files",
+                    uploadId, fileList.size(), fileNameSet.size() - 1));
+        }
```

Comment:
I know you're going to clean up later but it seems like a method that returns this JsonNode would encapsulate the above stanza.